sdks/python/hatchet_sdk/clients/durable_event_listener.py

```diff
@@ -0,0 +1,318 @@
+
+
+class _Subscription:
+    def __init__(self, id: int, task_id: str):
+        self.id = id
+        self.task_id = task_id
+        self.queue: asyncio.Queue[DurableEvent | None] = asyncio.Queue()
+
+    async def __aiter__(self) -> "_Subscription":
+        return self
+
+    async def __anext__(self) -> DurableEvent | None:
+        return await self.queue.get()
+
+    async def get(self) -> DurableEvent:
+        event = await self.queue.get()
+
+        if event is None:
+            raise StopAsyncIteration
+
+        return event
+
+    async def put(self, item: DurableEvent) -> None:
+        await self.queue.put(item)
+
+    async def close(self) -> None:
+        await self.queue.put(None)
+
+
+class RegisterDurableEventRequest(BaseModel):
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    task_id: str
+    signal_key: str
+    conditions: list[SleepCondition | UserEventCondition]
+
+    def to_proto(self) -> RegisterDurableEventRequestProto:
+        return RegisterDurableEventRequestProto(
+            task_id=self.task_id,
+            signal_key=self.signal_key,
+            conditions=DurableEventListenerConditions(
+                sleep_conditions=[
+                    c.to_pb() for c in self.conditions if isinstance(c, SleepCondition)
+                ],
+                user_event_conditions=[
+                    c.to_pb()
+                    for c in self.conditions
+                    if isinstance(c, UserEventCondition)
+                ],
+            ),
+        )
+
+
+class DurableEventListener:
+    def __init__(self, config: ClientConfig):
+        try:
+            asyncio.get_running_loop()
+        except RuntimeError:
+            loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(loop)
+
+        conn = new_conn(config, True)
+        self.client = V1DispatcherStub(conn)  # type: ignore[no-untyped-call]
+        self.token = config.token
+        self.config = config
+
+        # list of all active subscriptions, mapping from a subscription id to a task id and signal key
+        self.subscriptions_to_task_id_signal_key: dict[int, tuple[str, str]] = {}
+
+        # task id-signal key tuples mapped to an array of subscription ids
+        self.task_id_signal_key_to_subscriptions: dict[tuple[str, str], list[int]] = {}
+
+        self.subscription_counter: int = 0
+        self.subscription_counter_lock: asyncio.Lock = asyncio.Lock()
+
+        self.requests: asyncio.Queue[ListenForDurableEventRequest | int] = (
+            asyncio.Queue()
+        )
+
+        self.listener: (
+            grpc.aio.UnaryStreamCall[ListenForDurableEventRequest, DurableEvent] | None
+        ) = None
+        self.listener_task: asyncio.Task[None] | None = None
+
+        self.curr_requester: int = 0
+
+        self.events: dict[int, _Subscription] = {}
+
+        self.interrupter: asyncio.Task[None] | None = None
+
+    async def _interrupter(self) -> None:
```

Comment:
Is interrupter actually called? I think this is a relic of the action listener (which needs an external interrupt signal) but I don't think this does?

Reply:
okay yeah I don't think we're using the interrupt anywhere, but I'm not going to mess with this now - it's referenced in a handful of places